From 8e106917f409e3624f2bafd6ff90b99f23b25554 Mon Sep 17 00:00:00 2001
From: Kazuho Oku <kazuhooku@gmail.com>
Date: Fri, 20 Oct 2017 12:42:29 +0900
Subject: [PATCH] Merge pull request #1463 with tweaks

Do not send duplicate `upgrade` header when swithing to websocket
---
 lib/core/token_table.h | 2 +-
 misc/tokens.pl         | 2 +-
 t/40websocket.t        | 1 +
 3 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/lib/core/token_table.h b/lib/core/token_table.h
index 50df735c7..08f88583e 100644
--- a/lib/core/token_table.h
+++ b/lib/core/token_table.h
@@ -74,7 +74,7 @@ h2o_token_t h2o__tokens[] = {{{H2O_STRLIT(":authority")}, 1, 0, 0, 0, 0, 0},
                              {{H2O_STRLIT("strict-transport-security")}, 56, 0, 0, 0, 0, 0},
                              {{H2O_STRLIT("te")}, 0, 1, 0, 0, 1, 0},
                              {{H2O_STRLIT("transfer-encoding")}, 57, 1, 1, 1, 1, 0},
-                             {{H2O_STRLIT("upgrade")}, 0, 1, 0, 1, 1, 0},
+                             {{H2O_STRLIT("upgrade")}, 0, 1, 1, 1, 1, 0},
                              {{H2O_STRLIT("user-agent")}, 58, 0, 0, 0, 0, 1},
                              {{H2O_STRLIT("vary")}, 59, 0, 0, 0, 0, 0},
                              {{H2O_STRLIT("via")}, 60, 0, 0, 0, 0, 0},
diff --git a/misc/tokens.pl b/misc/tokens.pl
index fdff64a3d..9af1389c6 100755
--- a/misc/tokens.pl
+++ b/misc/tokens.pl
@@ -223,7 +223,7 @@ sub normalize_name {
 61 0 0 0 0 0 www-authenticate
 0 1 1 0 1 0 connection
 0 0 0 0 0 0 x-reproxy-url
-0 1 0 1 1 0 upgrade
+0 1 1 1 1 0 upgrade
 0 1 0 0 1 0 http2-settings
 0 1 0 0 1 0 te
 0 1 1 0 0 0 keep-alive
diff --git a/t/40websocket.t b/t/40websocket.t
index db3e65f4f..a321c034a 100644
--- a/t/40websocket.t
+++ b/t/40websocket.t
@@ -59,6 +59,7 @@ sub doit {
     like $rbuf, qr{^HTTP\/1\.1 101 }is;
     like $rbuf, qr{\r\n\r\n$}is;
     like $rbuf, qr{\r\nupgrade: websocket\r\n}is;
+    unlike $rbuf, qr{\r\nupgrade:.*\r\nupgrade:}is;
     like $rbuf, qr{\r\nsec-websocket-accept: .*\r\n}is;
     for my $i (1..10) {
         my $msg = "hello world $i\n";
